### PR TITLE
Add GetPropertyAsEnum

### PR DESCRIPTION
**Pull request by pepone**
This small PR implements `GetPropertyAsEnum` to read Ice properties as enumerated values

### CODE REVIEW

**Comment by pepone** [comment]
> Looks good to me. Why disallow numerics?

@bernardnormier suggested we do that because using numeric values is less readable

File: csharp/src/Ice/Communicator-Properties.cs [hunk 1]
```diff
@@ -52,6 +52,44 @@ internal PropertyValue(string v, bool u)
         public bool? GetPropertyAsBool(string name) =>
             GetPropertyAsInt(name) is int intValue ? intValue > 0 : (bool?)null;
 
+        /// <summary>Gets the value of a property as an enumerated type, the conversion does a case insensitive
+        /// comparison of the property value with the enumerated constants names and returns the matching enum or
```
Comment (bernardnormier) [inline csharp/src/Ice/Communicator-Properties.cs:56]: "enumerated constants names" are called enumerators.

File: csharp/src/Ice/Communicator-Properties.cs [hunk 2]
```diff
@@ -52,6 +52,44 @@ internal PropertyValue(string v, bool u)
...
+            lock (_properties)
+            {
+                if (_properties.TryGetValue(name, out PropertyValue? pv))
+                {
+                    try
+                    {
+                        pv.Used = true;
+                        if (int.TryParse(pv.Val, out int _))
+                        {
+                            throw new FormatException($"numeric values are not accepted for enumeratorion properties");
```
Comment (bernardnormier) [inline csharp/src/Ice/Communicator-Properties.cs:75]: I would use a simple "enum properties".

File: csharp/src/Ice/Communicator-Properties.cs [hunk 3]
```diff
@@ -52,6 +52,44 @@ internal PropertyValue(string v, bool u)
...
+                        }
+                        else
+                        {
+                            return Enum.Parse<TEnum>(pv.Val, true);
```
Comment (bernardnormier) [inline csharp/src/Ice/Communicator-Properties.cs:79]: I would use the argument name in front of true for readability.

File: csharp/src/Ice/Communicator-Properties.cs [hunk 4]
```diff
@@ -52,6 +52,44 @@ internal PropertyValue(string v, bool u)
...
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        throw new InvalidConfigurationException(
+                            $"the value `{pv.Val}' of property `{name}' is not valid for {typeof(TEnum)} enumeration",
```
Comment (bernardnormier) [inline csharp/src/Ice/Communicator-Properties.cs:85]: Suggested message:
the value `{pv.Val}' of property `{name}'  does not match any enumerator of {typeof(TEnum)}

File: csharp/test/Ice/acm/TestI.cs [hunk 5]
```diff
@@ -9,8 +9,7 @@ namespace ZeroC.Ice.Test.ACM
 {
     public class RemoteCommunicator : IRemoteCommunicator
     {
-        public IRemoteObjectAdapterPrx
-        createObjectAdapter(int timeout, int close, int heartbeat, Current current)
+        public IRemoteObjectAdapterPrx createObjectAdapter(int timeout, int? close, int? heartbeat, Current current)
```
Comment (bernardnormier) [inline csharp/test/Ice/acm/TestI.cs:12]: I think using `string?` for close and heartbeat would make more sense.

File: csharp/src/Ice/Communicator-Properties.cs (hunk 1 above)
Comment (pepone) [reply csharp/src/Ice/Communicator-Properties.cs]: fixed

File: csharp/src/Ice/Communicator-Properties.cs (hunk 2 above)
Comment (pepone) [reply csharp/src/Ice/Communicator-Properties.cs]: fixed

File: csharp/src/Ice/Communicator-Properties.cs (hunk 3 above)
Comment (pepone) [reply csharp/src/Ice/Communicator-Properties.cs]: fixed

File: csharp/src/Ice/Communicator-Properties.cs (hunk 4 above)
Comment (pepone) [reply csharp/src/Ice/Communicator-Properties.cs]: fixed

File: csharp/test/Ice/acm/TestI.cs (hunk 5 above)
Comment (pepone) [reply csharp/test/Ice/acm/TestI.cs]: fixed

**Pull request merged**